### PR TITLE
WasiP1 Support (only a subset for now)

### PR DESCRIPTION
**Pull request by bhelx**
I've still got some work to do here. But opening up now. API looks like this.

```java
var fakeStdin = new ByteArrayInputStream("Benjamin".getBytes());
var wasiOpts = WasiOptions.builder().setStdout(System.out).setStdin(fakeStdin);
var wasi = new WasiP1(wasiOpts);
var imports = new HostImports(wasi.toHostFunctions());
var instance = Module.build(new File("src/test/resources/compiled/greet-wasi.rs.wasm"))
                        .instantiate(imports);
var run = instance.getExport("_start");
run.apply();
```

Outputs can be set as PrintStreams and stdin can be passed an InputStream. You can also inherit from the parent process with: `WasiOptions.build().inheritSystem()`.

### CODE REVIEW

File: runtime/src/main/java/com/dylibso/chicory/runtime/WasiOptions.java
Comment (andreaTP) [inline runtime/src/main/java/com/dylibso/chicory/runtime/WasiOptions.java:6]: an idiomatic Java builder pattern will look like this (from the user's perspective):

```java
WasiOptionsBuilder.withStdin(mystdin).withStdout(mystdout).build();
```

**Review by andreaTP — COMMENTED** [review]
A couple of comments on the implementation.
The more I think about this the more I'm inclined to have a separate `wasi` module, wdyt?

**Comment by andreaTP** [comment]
Feel free to ping me for assistance if you have any trouble with Maven :+1: happy to help!

**Comment by bhelx** [comment]
> Feel free to ping me for assistance if you have any trouble with Maven 👍 happy to help!

be careful what you wish for 😈

File: runtime/src/main/java/com/dylibso/chicory/runtime/WasiOptions.java
Comment (bhelx) [reply runtime/src/main/java/com/dylibso/chicory/runtime/WasiOptions.java:6]: Is the pattern internally to just have a static class that looks exactly the same as the regular class? seems pretty wacky. https://www.digitalocean.com/community/tutorials/builder-design-pattern-in-java

Is there a better way?

File: runtime/src/main/java/com/dylibso/chicory/runtime/WasiP1.java
Comment (bhelx) [reply runtime/src/main/java/com/dylibso/chicory/runtime/WasiP1.java]: I'm gonna pick this up later. Since it's only two methods i feel like it's not worth doing this just yet.

File: runtime/src/main/java/com/dylibso/chicory/runtime/WasiOptions.java
Comment (andreaTP) [reply runtime/src/main/java/com/dylibso/chicory/runtime/WasiOptions.java]: It's the same thing but with mutable fields ... no magic...

**Comment by bhelx** [comment]
I moved it to it's own package. Eventually we should move this to it's own module. Maybe have it's own unit tests and stuff. it's a bit too much for me at the moment. I want to just use this to more easily test the runtime.

File: compilation-support-image/compile-rust-wasi.sh
Comment (bhelx) [inline compilation-support-image/compile-rust-wasi.sh:8]: I couldn't actually get this to work on my machine. The docker image is x86, so i made my own. but for some reason it's failing to create these tmp files. It gets to this point and then says that there is no file. It must be failing to run line 5.

File: runtime/src/main/java/com/dylibso/chicory/runtime/WasiOptions.java
Comment (bhelx) [reply runtime/src/main/java/com/dylibso/chicory/runtime/WasiOptions.java]: I'm confused, i still need both a builder class and a normal class with the same instance variables and setters right?

File: runtime/src/main/java/com/dylibso/chicory/runtime/wasi/WasiP1.java
Comment (bhelx) [inline runtime/src/main/java/com/dylibso/chicory/runtime/wasi/WasiP1.java:32]: I think this is sufficient to warn people for now. Not gonna document it just yet either.

File: runtime/src/main/java/com/dylibso/chicory/runtime/wasi/WasiP1.java
Comment (bhelx) [inline runtime/src/main/java/com/dylibso/chicory/runtime/wasi/WasiP1.java:38]: I stubbed out as many WASI p1 functions as i could find. Oftentimes they are imported and never called. But if they are, it should trap here with the name of the wasi function that is not implemented.

The only real implementations at the moment are fd_read and fd_write and only for file descriptors 0,1, and 2

File: runtime/src/main/java/com/dylibso/chicory/runtime/wasi/WasiP1.java
Comment (bhelx) [inline runtime/src/main/java/com/dylibso/chicory/runtime/wasi/WasiP1.java:307]: In the future we'll probably keep a map of open files keyed by their fd and do a lookup here. that will require some of the directory mapping stuff which i want to avoid.

File: runtime/src/main/java/com/dylibso/chicory/runtime/WasiOptions.java
Comment (andreaTP) [reply runtime/src/main/java/com/dylibso/chicory/runtime/WasiOptions.java]: > instance variables and setters

you need only the constructor and the getters

File: compilation-support-image/compile-rust-wasi.sh
Comment (andreaTP) [reply compilation-support-image/compile-rust-wasi.sh:8]: That's wired, thanks for reporting!
Unfortunately, I don't have an ARM machine to test, so I'll need someone's help to debug this one.

**Review by andreaTP — CHANGES_REQUESTED** [review]
LGTM after merging #166 in this branch, and let's not forget to take a closer look at the compilation issue.

File: runtime/src/main/java/com/dylibso/chicory/runtime/WasiOptions.java
Comment (andreaTP) [reply runtime/src/main/java/com/dylibso/chicory/runtime/WasiOptions.java]: resolved in #166

**Comment by andreaTP** [comment]
Another follow-up task is gonna be documentation, updating the readme etc., not required in this PR as it's already big enough.

**Review by andreaTP — APPROVED** [review]
LGTM, we would need to iterate on the compilation/scripts etc. but that's fine and doesn't hurt.

**Comment by bhelx** [comment]
> Another follow-up task is gonna be documentation, updating the readme etc., not required in this PR as it's already big enough.

Agreed. I think let's hold off until we're more confident though. I'd prefer for this to be hidden for the time being. I just want to use it to get some more modules in here for testing purposes.

File: compilation-support-image/compile-rust-wasi.sh
Comment (bhelx) [reply compilation-support-image/compile-rust-wasi.sh:8]: well, docker images are ISA specific. But i may be able to detect if this is running in mac and force it to emulate x86. Will try some things out don't worry about it for now.

**Pull request merged**
